Custom user script directories

mastep84 opened the pull request:
Dear Sebastian,

we discussed this topic a while ago and recently I have taken a shot at extending ProMesh to support the addition and removal of custom user script directories for monitoring -- analogously to the default directory ./promesh/scripts. To this end, the user-specified directories are stored in ./promesh/custom_user_script_dirs. 

In this context, it would be generally useful to facilitate referring to additional data, e.g. raster files, using relative paths in scripts that are executed from within ProMesh, but I haven't been able to resolve this issue yet.

Best,
Martin

mastep84 committed:
Extended the scripts dialog to support the addition and removal of a custom user script folder.

mastep84 committed:
The previously introduced feature 'addition/removal of a single custom user script folder' has now been further generalized to support multiple folders.

sreiter commented on src/app.cpp:
```diff
@@ -90,6 +90,42 @@ QDir UserScriptDir()
+	std::vector<QDir> vCustomUserScriptDirs;
+
+	if(UserDataDir().exists("custom_user_script_dirs")){
+		QFile inFile((app::UserDataDir().absoluteFilePath("custom_user_script_dirs")));
```
Instead of maintaining a custom file and parsing lines from that, it would probably much more convenient to simply use Qt's `QSettings` object and its `value(...)` method with a key "customUserScriptPaths" and a `QStringList` value. Each string in that list corresponding to a user path:

```
QSettings settings;
QStringList userPaths = settings.value ("customUserScriptPaths").toStringList ();
```

sreiter commented on src/main.cpp:
```diff
@@ -116,7 +116,7 @@ int main(int argc, char *argv[])
 	QCoreApplication::setOrganizationName("ProMesh");
     QCoreApplication::setOrganizationDomain("promesh3d.com");
-    QCoreApplication::setApplicationName("ProMesh4.3.10");
+    QCoreApplication::setApplicationName("ProMesh4.3.18");
```
While 4.3.18 is more accurate, changing this string will create a new file where the QSettings contents are stored. When a user opens ProMesh with the changed application name, old settings will be gone. Currently this only affects the file open/save path. But in the future it would also apply to the custom script paths, if you will adjust their storage to QSettings.
Maybe you could simply change it to "ProMesh"?

sreiter commented on src/modules/mesh_module.cpp:
```diff
@@ -289,6 +301,37 @@ void MeshModule::browseUserScripts()
+	                                                | QFileDialog::DontResolveSymlinks);
+
+	if(!customUserScriptDir.isEmpty()){
+		QFile outFile(app::UserDataDir().absoluteFilePath("custom_user_script_dirs"));
```
If you go the QSettings route, you could simply write:

```
QSettings settings;
QStringList userPaths = settings.value ("customUserScriptPaths").toStringList ();
userPaths.push_back (customUserScriptDir);
settings.setValue ("customUserScriptPaths", userPaths);
```

sreiter commented on src/tools/script_tools.cpp:
```diff
@@ -285,6 +285,16 @@ void RegisterScriptTools(ToolManager* toolMgr)
 	ParseDirAndCreateTools(toolMgr, QDir(":/scripts"), "Scripts", shell);
 	ParseDirAndCreateTools(toolMgr, app::UserScriptDir(), "Scripts", shell);
+
+	if(app::UserDataDir().exists("custom_user_script_dirs")){
```
This check is not really necessary. And the location where the customUserScriptDirs are stored is actually an implementation detail of `app::CustomUserScriptDirs()`, I think.

sreiter commented on src/tools/script_tools.cpp:
```diff
@@ -317,8 +327,18 @@ int RefreshScriptTools(ToolManager* toolMgr)
 	ParseDirAndCreateTools(toolMgr, QDir(":/scripts"), "Scripts", luaShell, true);
 	ParseDirAndCreateTools(toolMgr, app::UserScriptDir(), "Scripts", luaShell, true);
 
+	if(app::UserDataDir().exists("custom_user_script_dirs")){
```
This looks very similar to the block above (as do the two lines above). Maybe there could be a separate function which does the parsing of standard and user tools? E.g.:

```
void ParseAndCreateTools (toolMgr, luaShell, false/true);
```

I hope the idea is clear. If not please tell me 😄

sreiter reviewed (commented):
Hey Martin,

thanks a lot that you took the time to implement this feature. This is a very cool addition! It makes some stuff much more simple, like having some promesh scripts in a ug4 app or separate github repos with cool scripts :)

I added some comments - mainly to encourage you to use QSettings instead of a custom file to store those user paths. Mainly because it results in less code, easier maintenance and less potential bugs and issues.

Did you have the chance to check what happens if you have the same script with the same name in multiple user paths? Would this result in multiple tools? Or would there even be an error? I think it may be handled just fine. I was just wondering... 😃 

I agree that relative paths could be a cool addition. We should probably talk about this for a separate pull request.

Please let me know what you think of my suggestions.
Thanks again!

mastep84 committed:
Modified custom user script support to use the QSettings object for storing user paths instead of a custom file.

mastep84 commented:
Hi Sebastian,

thank you very much for your thorough review and constructive feedback. I modified the implementation according to your suggestion to use the QSettings object for storing user paths instead of a custom file.

Regarding your question about identical scripts in multiple user paths: As you suspected, these actually result in multiple tools listed under the corresponding group name that is determined by each user directory.

In script_tools.cpp I removed the if-check that you highlighted as it obviously does not apply anymore, though I refrained from introducing a separate function which does the parsing of standard and user tools.

Best regards,
Martin

P.S. I took your advice and changed line 119 in main.cpp to "QCoreApplication::setApplicationName("ProMesh");"

mastep84 committed:
Modified .gitignore for Visual Studio Code.

sreiter commented:
Thanks Martin!
